addons/package_hierarchy/models/stock_quant_package.py

```diff
@@ -0,0 +1,223 @@
+"""Packages with inheritance."""
+
+import logging
+from itertools import chain
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+from odoo.tools.float_utils import float_is_zero, float_compare
+
+_logger = logging.getLogger(__name__)
+
+
+class QuantPackage(models.Model):
+    """ Add the ability for a package to contain another package """
+
+    _inherit = "stock.quant.package"
+    _parent_name = "parent_id"
+    _parent_order = "name"
+    _order = "id"
+
+    display_name = fields.Char("Display Name", compute="_compute_display_name")
+    parent_id = fields.Many2one(
+        "stock.quant.package",
+        "Parent Package",
+        ondelete="restrict",
+        help="The package containing this item",
+    )
+    top_parent_id = fields.Many2one(
+        "stock.quant.package", compute="_compute_top_parent_id", store=True
+    )
+
+    children_quant_ids = fields.One2many(
+        "stock.quant", string="All content", compute="_compute_children_quant_ids"
+    )
+    children_ids = fields.One2many("stock.quant.package", "parent_id", "Contained Packages")
+    depth = fields.Integer(compute="_compute_depth", store=True)
+
+    @api.depends("parent_id", "parent_id.top_parent_id")
+    def _compute_top_parent_id(self):
+        for pack in self:
+            parent = pack.parent_id
+            if not parent:
+                pack.top_parent_id = False
+            elif parent.top_parent_id:
+                pack.top_parent_id = parent.top_parent_id
+            else:
+                pack.top_parent_id = parent
+
+    @api.constrains("parent_id", "children_ids")
+    @api.onchange("parent_id", "children_ids")
+    def _constrain_depth(self):
+        max_package_depth = self.env.user.get_user_warehouse().u_max_package_depth
+        for pack in self:
+            top_parent = pack.top_parent_id or pack.parent_id
```

Comment:
Is there any chance where this will happen? If it has parent_id we should have top_parent_id, doesn't it?